### ISSUE TITLE
[Discussion] Solving Search in Ghost?

### ISSUE THREAD

**Issue by ErisDS** [TryGhost/Ghost#5321]
The issue to add [Search to the Posts API](https://github.com/TryGhost/Ghost/issues/306), is literally the oldest open issue in this repository. It’s a specialist subject, so we’ve always hoped someone with specialist knowledge would come forward to help us solve it, but unfortunately that hasn’t happened.

To try to increase visibility of the problem we set up a page advertising [contributor roles](https://ghost.org/contribute/search-lead/) - detailing that we were looking for someone to help solve adding search to Ghost. That didn’t work either :(

Unfortunately, adding search is a particularly complex challenge. This is another attempt to try to get engagement from the wider OSS community on how to do this - we’d really appreciate the opinions and input of anyone who knows their search-fu.

---
## Use cases

There are two key use cases for search inside of Ghost itself - one being in the admin panel to find a post, tag, user or perhaps even a setting that you want to change. This would make a significant improvement to the usability the admin interface. 

The other is adding search to the frontend of a blog, so that it becomes easier to find content. The main things that need to be searchable are post titles and content, tag names and description, user names, and probably meta titles and descriptions for the different resources. There are plenty of use cases where other things would also need to be searchable, so this will need to be extensible, but these are the key fields.
## Approaches

The best way to make both of these use cases possible, as well as making the search feature available to any apps or other extensions of Ghost in future, is to make it available through the API. 

There are generally two approaches to solving search that would work in Ghost that I am aware of - the first being to use the FTS features of SQL and the second being to use a third party search tool of some description. 

The upside to FTS is that, if we get it working well, it could work for all Ghost installs without any need to install additional (and likely complex) dependencies. The downside is that getting FTS to work for all three of sqlite, MySQL and PG is a sizeable challenge & the [status of FTS in knex is unknown](https://github.com/tgriesser/knex/issues/203) - but perhaps it can be done as a bookshelf/knex plugin, in a way that will benefit the wider community as well?

The upside to using a 3rd party tool, is that there are many modern and advanced systems that we could leverage to make an exceptional search feature. The downside is that most of them require additional complex dependencies* to be installed, which mean they aren’t viable options for Ghost core. 

One of the tools that has been recommended is lunr.js, and there is a [plugin for Ghost themes](https://github.com/jamalneufeld/ghostHunter) which combines lunr & the RSS feed to make search possible on the frontend. Supposedly lunr also works in node, not just the browser, but I’ve not found much information about it.

Other tools that are popular for search include lucene and elasticsearch and there are also numerous modules for node built on top of leveldb and redis databases. There are likely solutions out there that I haven’t heard about, so I’d be interested to hear ideas from others.

\* complex dependencies includes binaries or external services - Ghost is currently installable on almost any platform without too much fiddling with compilers etc, and we want to keep it that way!
## Moving forward

The ideal solution, in my mind, looks something like:
- Have a rudimentary version of search built into Ghost via the API - using FTS or something else - that will work for 90% of blogs (i.e. relatively small amounts of content).
- Implement this version of search to be easily extensible, so that it’s very easy to create an app that replaces the search with something more heavy-duty like elastic search or lucene, to support larger blogs. 
## The immediate future

As part of Zelda, we want to introduce a search bar to the admin interface which provides auto-complete style results for posts, tags and users. This should be possible using our existing API, some minor modifications (to allow us to fetch certain fields) and Ember. It’s not ideal, but I think something incredibly rudimentary is better than nothing at all! This will be spec’d in a separate issue.
## Continuing this discussion

I’m really, really looking for the community to get involved and share their ideas here. This is a relatively specialist subject and I know there are tonnes of developers out there that have a great deal more experience implementing search than I do. 

I’d like to hear thoughts, ideas and experiences on FTS - will we be able to make it work for all 3 databases without having to write too much custom code? Could a bookshelf plugin work? Is it an absolute nightmare not worth pursuing?

What about lunr? Is it a viable server-side option? Does it have too many requirements or not work for large data sets, or is it the perfect basic option provided it can easily be overridden with elastic search or something else?

Is there another solution that isn’t mentioned here? Remember we can’t have leveldb or redis or other complex dependencies in core and it has to work across sqlite3, mysql and postgres.

And what about the API? What would a good search API look like? Who out there has a great RESTful search API that we should take inspiration from?
If you know someone you think could answer any of these questions, please link them here and ask them to get involved!

Thanks :)

**Comment by riyadhalnur** [comment]
I haven't seen any viable option out there yet except lunr. This look promising [norch](https://github.com/fergiemcdowall/norch) but the documentation needs more work. Norch follows almost the same syntax as lunr; so for anyone to start with it should be fairly easy. My 2 cents

**Comment by ErisDS** [comment]
@riyadhalnur norch fails the requirement of not needing complex dependencies.

**Comment by CWSpear** [comment]
@nichcurtis is a search master. He says Sphinx fits your bill and has an npm module, but he hasn't used it with node before.

Pinging him to try and get him involved!

**Comment by olivernn** [comment]
Lunr should work just as well in node as it does in the browser. It doesn't have any dependencies, so should be straightforward for ghost users to set up.

I don't know the scale of the indexing/searching you are trying to achieve but lunr can scale to fairly large corpus sizes, though it is never going to compete with the likes of Solr, ElasticSearch etc. I'm more than happy to try and help with any load testing you might want/need to do though.

Let me know if I can be of any help helping you decide, even if it is to say that lunr isn't the right fit.

**Comment by CWSpear** [comment]
Lunr's scalability would be directly related to memory, correct? Isn't it all stored in memory when it does the searches? 100MB worth of data is a decent sized blog, but not unfeasible, and since this app doesn't normally have a huge need of memory (I'm running my blog on a 500 MB VM), it could definitely be a problem, no?

**Comment by ErisDS** [comment]
I imagine that for very large blogs, the memory footprint could get problematic, but whether or not this is a serious issue for 'most blogs' would depend very much on the relationship between amount of text & the size of lunr's footprint. @olivernn some sort of testing around lunr.js memory footprints for an average blog would be really helpful here I think - I'm sure we could muster up some anonymised stats on what an average blog looks like if needed. 

In terms of solving this elegantly, an idea we're working towards at the moment is to start delivering 'internal apps' - where bits of new functionality are delivered as an 'app' which is automatically installed and either enabled or disabled. Search could be delivered this way, as an app which is always present but disabled by default.

The benefit to this is that not everyone wants or needs search for their blog - so if you're happy without it, you can leave it disabled. If you want it you can enable it, and if it's too memory intensive on your blog you can (in theory) swap it out for the 'ElasticSearch' app instead. There are a lot of puzzle pieces missing for apps still, but it's worth knowing this is the way we intend to move forward.

**Comment by olivernn** [comment]
@CWSpear lunr is an in memory search index, and its memory usage is going to be related to the number and size of documents being indexed, so memory use may be an issue for larger blogs.

@ErisDS I can put together some benchmarks of lunr's memory footprint if you can provide some representative sample data. It will at least give you a data point to help you make the decision.

I don't know Ghost _at all_ but from what I see it is using SQLite as a datastore? Doesn't SQLite have a full text search index that could be used here? Or am I completely misunderstanding the architecture of Ghost.

**Comment by halfdan** [comment]
@olivernn Ghost uses SQLite3 by default, but also supports PostgreSQL and MySQL. As @ErisDS described we would need a solution that can cover all three databases.

**Comment by leonkyr** [comment]
I would go for ElasticSearch, why:
- it platform independent (you only need java)
- you can get deploy an archive together with Ghost, unzip and start the node
- it has rest api, node.js sdk and scale well

The good idea would be to implement a microservice for search that can (re)index documents and search them. Inside it can be ES or FTS. ES supports text highlighting and a lot of other things.

Btw, from this thread it is not clear what is the final choice.

**Comment by kowsheek** [comment]
:+1: on the ElasticSearch suggestion. It has a huge community which is super active. Lots of hosting support as well. ES would work nicely for other languages as well.

Ideally, ES should come packaged, but give the option to use a built in node or one that's started elsewhere (or none at all).

Ditto what @leonkyr & @ErisDS says, especially since PostgreSQL has a pretty awesome FTS, the searching shouldn't be limited to ES, but may be the default.
> The good idea would be to implement a microservice for search that can (re)index documents and search them. Inside it can be ES or FTS. ES supports text highlighting and a lot of other things.

**Comment by ErisDS** [comment]
> it platform independent (you only need java) 

> Ideally, ES should come packaged

ElasticSearch would be awesome, but I'm really struggling to see how we could use it in Ghost in a way that didn't make it impossible to install?

**Comment by CWSpear** [comment]
@ErisDS I think your *only* option is Lunr (or something similar to it), because everything else at the *very least* requires a persistence layer (i.e. https://github.com/tj/reds) or a backing process (i.e. elasticsearch) (or both). *Or* they are specific to a particular database (https://dev.mysql.com/doc/refman/5.0/en/fulltext-natural-language.html).

So we need something that wraps around the the different databases you support/plan on supporting (SQLite, Postgres, MySQL) with a common API, or something that puts everything it in memory like Lunr.

**Comment by riyadhalnur** [comment]
ElasticSearch as good as it may be, IMHO I don't see it as a possible solution for Ghost. To have to install Java to make it run for a small application is a big hassle among other things.

why couldn't we just use the indexing capabilities of the DBs Ghost supports and write a high level API for it so developers and users (there are some who will) can use it however they see fit? in other words, a polyglot framework. it's a lot of work but the payoffs will be immense.

**Comment by ErisDS** [comment]
I agree - Java cannot be a dependency for Ghost. 

I also agree about a high level API - the way I imagine it would be that Ghost would offer a `/api/search?q=` endpoint from it's API, but how that search gets performed would be configurable. The underlying structure would be ideally to have a default that is built in to Ghost and then be able to plug in ES or Solr or whatever takes your fancy using an app.

What we need (and the intention of looking for a Search Lead) is to figure out how that will work. Firstly, which is the default? I think either lunr or FTS could be the default, but it needs someone familiar with implementing this sort of thing to do a bit of a test and spec it out. Second of all, how does the default work. Thirdly, what does the  fine details of the API endpoint/ look like and finally how can an app hook in and override the search functionality whilst keeping the API the same?

**Comment by riyadhalnur** [comment]
ES and Lunr.js use JSON stores. i would see the system as generating JSON stores which an app can use (hook into) to power the API

**Comment by nagug** [comment]
Just a wild thought.. 
Can the raw queries of knex be used and a fts performed? Some amount of overhead would be there to form the query depending on the db used. 

Not sure an app approach would help.. But app approach might introduce security issues. 
Thanks

**Comment by codeincarnate** [comment]
This thread has died down, but I wanted to hop in because it's crucial that Ghost gets this right.  I'm going to talk about why, and then outline an approach that Ghost can take to avoid this.

I've implemented search on many Drupal based sites and it's easy to mess up.  Way back, Drupal devs decided to implement a search solution on top of MySql and it was plagued with problems.  The search results were ineffective and performance was terrible.  It became common knowledge to simply [disable it altogether](http://www.drupalcoder.com/blog/performance-tip-disable-drupals-core-search-indexer-when-using-apache-solr.html).  It was so bad that it was effectively unusable and the community adopted the use of [Solr](https://lucene.apache.org/solr/) in droves.

This is definitely undesirable, and I believe this outlines the following requirements:

- Search should be a pluggable backend.  The search requirements for a blog with 100 articles and one with thousands are very different.
- Search should be performant and relevant out of the box.  It's not enough to have something that kind of works, it should work well.
- Given the above, it should not be complicated to setup search.  By default it should require no setup.
- It should be a fairly straightforward to move from one search backend to another as needs change.

### Backends 

In regards to backends, there are a number that are available as a default choice:

- [Lunr.js](http://lunrjs.com/)
- SQL based Full Text Search
- [search-index](https://github.com/fergiemcdowall/search-index)
- [levi](https://github.com/cshum/levi)

Lunr.js is nice in that it has a relatively active community.  However, Lunr has a couple of downsides.  First, it's purely Javascript which could negatively effect performance.  Second, Lunr stores it's index by serializing it to JSON.  This would likely be slow, but also makes managing the index tricky.

Full Text Search (FTS) features of various databases could definitely be used and would likely have fairly solid performance.  The downside of this is that it would be relatively difficult to code and maintain.  It would effectively have to be implemented for MySql, Postgres, and Sqlite.  Another more subtle difference is that this also means that the database begins storing metadata of the blog in addition to the blog data directly.  This can make migrations and other tasks trickier since there's a portion of the database that could effectively be regenerated at any time.

Finally, search-index and levi.  Both are implemented on top of [LevelUp](https://github.com/Level/levelup) which is a LevelDB interface.  At first glance, this would appear to introduce a node-gyp dependency, however LevelUp has pluggable persistence.  This means that a solution built on top of LevelUp could use something like [MemDOWN](https://github.com/level/memdown) by default, and switch to something like [LevelDown](https://github.com/Level/leveldown) as performance needs increase.  

It could also be possible to do some work to use [node-pre-gyp](https://github.com/mapbox/node-pre-gyp) for LevelDown so that there's no need for users to build binaries at all.  In fact, this is already the case for Sqlite support in Ghost!  This would also have the added advantage that all index data would be separate from the blog data itself.

Personally, I feel the LevelUp backed path is the way to go.  It's an approach that allows for very straight-forward setup, is likely to have solid performance, is pretty full-featured, and avoids maintaining separate search backends (in the SQL FTS case).

### Indexing

Thanks to bookshelf, it should be straightforward to hook into events that are occurring at the model layer.  This is [already being done](https://github.com/TryGhost/Ghost/blob/7761873db7689f83b7f3569e4c28624cd750f6c7/core/server/models/base/index.js#L64) and could be extended to meet indexing needs.  There would likely need to be extra added to do index clearing and re-indexing as necessary.


### Pluggable Backends

I'm not aware of everything that goes into Ghost apps, but it would be a good avenue for other backends.  I'm sure this would be important for Ghost Pro and larger blogs so that a solution like Solr or Elastic can be used in that case.

I'm part way through doing a _very_ barebones implementation of the above.  Regardless of direction, I'd love to hear some feedback and thoughts.

**Comment by codeincarnate** [comment]
An update to my last comment, I have a _super, super_ basic implementation [here](https://github.com/codeincarnate/Ghost/tree/search-basic).  This doesn't have pluggable backends or anything, just some indexing and the ability to pass in queries at the moment.

**Comment by CWSpear** [comment]
@codeincarnate looks like that's a private repo (or maybe just a bad link?) as I got a 404 when trying to visit it.

**Comment by codeincarnate** [comment]
@CWSpear Missing an "s" at the end, fixed now!

**Comment by gergelyke** [comment]
@ErisDS any update on this?

**Comment by codeincarnate** [comment]
@ErisDS would love to have some feedback.  My business partner is getting married this weekend, but after that I'm going to have more time to really take a crack at this.

**Comment by sethbrasile** [comment]
@codeincarnate if you get into this, feel free to throw some tasks my way if you'd like.

**Comment by codeincarnate** [comment]
@sethbrasile That's a very generous offer and I'll definitely do so. Hoping to get some feedback from @ErisDS before taking a deep dive into this.

**Comment by ErisDS** [comment]
@codeincarnate thanks for the enormous writeup, and apologies for taking a while to digest it.

It seems to me that, providing that we can have a non-binary/in-memory dependency as the default, this approach looks like it may be a great solution.

Having a node-pre-gyp wrapper for the binary version is very desirable, even if it's not the default. I would recommend that we avoid having another default dependency which requires a binary, even with node-pre-gyp, because it does still add overhead both to users installing and to developers (switching between node versions, which is becoming commonplace during testing, is harder with binary deps).

I'd be interested to see some tests to show at roughly what point in-memory search becomes untenable. E.g. how many 500 word posts is the limit, assuming a DO $5 droplet / 512MB ram? This could help us to provide useful documentation on how and when to upgrade to 'advanced' search, as well as perhaps even add warnings inside Ghost itself.

Not sure if you've seen this, but here's the documentation for how to configure a custom storage module, which is a similar 'pluggable' extension to Ghost: https://github.com/TryGhost/Ghost/wiki/Using-a-custom-storage-module. I'd imagine the extra config for search would be similar - an npm install & a config.js update. This could be done in a more managed way maybe using the app platform in future, but for the time being this approach works.

The only major question I'm left with atm is why [levi](https://github.com/cshum/levi) over [search-index](https://github.com/fergiemcdowall/search-index)?

**Comment by jberkus** [comment]
PostgreSQL Expert, JS neophyte here, just putting in:

For Postgres, at least, full text search can be added in ways which are invisible to the user, and which work fine with backup/restore.  Further, if you only support simple boolean operations (i.e. AND/OR words) and don't support more complex operators, it's trivially easy to write a library which would convert user-obvious syntax into the syntax supported by each FTS engine.  The one user-visible operation would be requiring the user to choose a language, as FTS dictionaries are language-specific, and the dictionary would become an optional installation requirement.

MySQL is more complicated, especially if you need to support core MySQL and not require MariaDB or Percona.  Core MySQL, AFAIK, does not support FTS for InnoDB tables, forcing you to make a rather substantial user-visible change which affects reliability.  This is why prior to Maria/Percona, most MySQL users used Spinx search instead.

**Comment by jberkus** [comment]
The main reason to support SQL-based FTS would be in order to support large websites with several GB of text to search.  Does anyone even use Ghost for that?

**Comment by codeincarnate** [comment]
@jberkus Awesome run down in terms of Postgres.  For the most part, many people are likely to be running Ghost on a small blog and there's probably not going to be more than 100mb of text or so and even that is probably an over-estimation.

From the perspective of ghost, it can run on top of MySql, Postgres, or Sqlite.  In terms of search, MySql is likely the one to hold it back.  More than that however, dealing with a wide array of potential database configurations is not ideal for a default search function across small blogs.  Keeping a solution which

As I mentioned above, this is the point of having search be a completely pluggable interface.  If you're running Ghost on top of Postgres it makes a lot of sense to also have search in Postgres as well.  For a service like Ghost Pro, running search on top of Elastic or Solr makes sense.  For the easy case, some zero configuration and no worries is definitely for the best.

**Comment by jberkus** [comment]
Yeah, seems like a good approach. If you make it pluggable, I might write the PostgreSQL backend.

**Comment by codeincarnate** [comment]
@ErisDS thanks for the feedback!

I'm going to look into the image storage and how all that works.  For reference, where is the code that implements that?

As far as levi vs search-index, I suppose I don't have a strong pull towards one or the other.  I'd mostly be interested in testing both and see how they fare both in terms of performance and in terms of search accuracy.  They're quite similar, so hopefully some action would spur the community towards adopting a solid option.

At the moment, [search index doesn't have a license](https://github.com/fergiemcdowall/search-index/issues/177) so that will need to be clarified at least.  Both of them are also explicitly requiring LevelDOWN so that will need to be changed as well.  

I'm going to start breaking out some issues and moving forward with this.

**Comment by CWSpear** [comment]
I believe there's been talk of dropping Postgres support, by the way.

Also, I don't think it's been discussed, and it may not be viable for this use-case, but I think it's definitely worth mentioning: there are search-as-a-service options available, such as [Algolia](https://www.algolia.com/). While some obvious cons, a major pro would be that (if you had an account), it'd work out of the box on a vast array of systems running Ghost (since all of those difficulties are abstracted away by being taking care of off-site).

**Comment by jberkus** [comment]
@CWSpear 

Yes, that's why I got involved.

**Comment by CWSpear** [comment]
I mean dropping Postgres altogether. No (official) support for Postgres anything for Ghost.

**Comment by jberkus** [comment]
Yes, I know, see #5878.

**Comment by halfdan** [comment]
@CWSpear The discussion is about dropping **official** support and leaving it up to the community to keep psql supported.

**Comment by CWSpear** [comment]
Alright, just making sure we were all aware =)

**Comment by jberkus** [comment]
<--- the community (plus, y'know, some other folks)

**Comment by mike182uk** [comment]
Hey guys, just thought i'd chime with something i had been playing about with: https://github.com/mike182uk/Ghost/pull/1 (raised a PR on my own fork, just to make it easier to see what is going on and what changes i've made).

I have basically taken the same approach discussed above with having pluggable backends (strategies). My example shows how posts can be searched using Lunr as a strategy.

You can search for a post by adding the `search` param to the querystring

```
http://<base_url>/ghost/api/v0.1/posts/?search=<query>
```

The index is updated whenever you add, edit or delete or post.

The search strategy is configureable via the config, so this should make it easier to extend and add other search strategies (i.e algolia). 

This is just a working PoC and would need more work (indexers etc.) but maybe of help to somebody :smiley:

**Comment by codeincarnate** [comment]
@mike182uk cool!  I've taken a look at the code and it's broadly similar to what I've been doing.  I'm not sure how good of an idea it is to call it a strategy, and there is definitely more that needs to be done in terms of loading the index on startup for instance.  I'll look at it more in depth over the weekend.

**Comment by jberkus** [comment]
What will the search queries look like?

**Comment by mike182uk** [comment]
@jberkus with the approach i've taken, the main query is altered to use `WHERE uuid IN (<uuid>, <uuid>, <uuid>)`. This means that an external system (i.e lunr, solr, algolia) would return the UUIDs of documents that matched the query, then these are passed through to the query that retrieves the data from the db

@codeincarnate I called them strategies as its the [strategy pattern](https://en.wikipedia.org/wiki/Strategy_pattern) that is being used

**Comment by jberkus** [comment]
@mike182uk I meant "what does the search query text look like"?  Rather than the follow-up query.

**Comment by jberkus** [comment]
I'm thinking in terms of how I would write a search plugin, so I'm curious what the search syntax looks like.

**Comment by mike182uk** [comment]
@jberkus ah i'm not sure, i don't think its been mentioned here. My implementation just takes a string from the query string (`http://<base_url>/ghost/api/v0.1/posts/?search=<query>`)

@codeincarnate I've added an indexer to my implementation now if you want to check it out. This means that it will be up to the indexer to know how to index models. In the case of my Lunr.js example, when the `indexModelCollection` method of the indexer is run, the indexer could load a disk saved Lunr.js index to save having to re-index everything when ghost is loaded. I would imagine this sort of thing would be different between strategies.

How does everyone feel about the approach i've taken? I would love to get some feedback (@erisds maybe?). I'd be happy to more time into this if i'm going in the right direction :smile: 

If anybody wants to test out my approach, just fork my fork, or git checkout my PR on my fork. Once you have started ghost, the post models will be indexed and you can query them via the API:

```
http http://<base_url>/ghost/api/v0.1/posts/?search=<search_query> 'Authorization:Bearer <auth_token>'
```

Adding new posts, updating or deleting existing posts will update the index.

**Comment by jberkus** [comment]
Right, I'm asking: what goes in place of the placeholder <search_query>?  What does that look like?

**Comment by ErisDS** [comment]
@codeincarnate code for storage modules lives here: https://github.com/TryGhost/Ghost/tree/master/core/server/storage. Issues that are worth reading for some context inc #4600 & #2852 - I like the idea of trying out both levi and search-index to see which is best. 

@mike182uk awesome that you have a similar implementation, I have taken a very quick look and it seems like all the pieces are plugging in in roughly the way I'd imagined. I'm interested to know why you picked lunr?

@jberkus I recommend taking a look at GQL: #5604 - which is used for filtering - for some inspiration for what a search query might look like. I'd imagine we'd want to do things a bit differently for search - but it's the only even related spec that exists in Ghost atm. In order to determine the actual syntax, we'd need to determine what the basic set of queries that most 'strategies' can support is, as well as consider how we could extend this out from posts to include more resources, and go from there to come up with a plan.

One big question I guess is what to do with the custom 'strategies' & how to manage configuring them - should it be exactly the same as 'storage' for now, e.g. use `/content/search/`? We don't have enough pieces to do this via an app just yet. 

Do any of these modules which use memory do any sort of reporting on how much memory they are using? I'm wondering about what we can do in terms of managing the potentially enormous memory usage jump we're going to get.

**Comment by jberkus** [comment]
The GQL thing is cool, I hadn't seen that.

However, it doesn't define a syntax for full-text searches.  Personally, I like to keep things simple, so I'd vote for a syntax which accepts:

     "brunch breakfast eggs"

which would be equivalent to:

     "brunch and breakfast and eggs"

REST searches would use this syntax:

     "brunch&breakfast&eggs"

Also support basic OR searches:

     "brunch or breakfast or eggs"
     "brunch|breakfast|eggs"

I don't think we want to support AND/OR groups, etc.  At least not right away.

**Comment by jpwynn** [comment]
I'd like to make a Modest Proposal (in the Jonathan Swift tradition), acting on the assumption there's still no meaningful progress on this issue. It is just so unbelievably bad to not have **any** search, not even a dumb keyword lookup, that you might consider something like the following:

* settle for simple case-insensitive keywords like 'heroku' to find all posts that include heroku in either the title or the body. Up to you whether "breakfast eggs" is a single term, or two terms ANDed because... and this to the main point... either way is better than nothing.
* ignore performance, ignore indexing, let the database do the work (for now) because 'slow-ish' is a LOT better than 'nonexistent'
* if there is not a database-independent abstraction for a simple 'ILIKE' sql query, hold your nose and implement the required handful of lines of database-specific code for the small number of databases you support.

Let "making search better" become the oldest open issue if circumstances and complications dictate, because it's a lot better than having no search.

IMO, you are letting 'ideal' be the mortal enemy of 'adequate for most bloggers'.

**Comment by jberkus** [comment]
FWIW, "ILIKE" should be database-independent, at least for the two supported and one quasi-supported database.

**Comment by mike182uk** [comment]
I've been away and not had chance to look at this recently. `0.7.2` made some changes that has required me to change my implementation slightly. I have changed how i am filtering posts to use the new filter functionality - basically you still pass a search query along the query string when making the http request, but this now gets converted into a filter. If there is already an existing filter, it is appended on to this. I've also made it so the indexer configuration is coming from the main config now. You can see my implementation [here](https://github.com/mike182uk/Ghost/pull/1/files) (feedback is welcome!)

@ErisDS as for why i chose lunr - its by far the easiest solution to integrate and has the least dependencies. The main concern with lunr is the memory footprint as the index is stored in memory. Its a bit hazy at what point that really becomes a problem (i.e for the average blog i don't see this being a problem).

I was trying to tackle the problem with a more pragmatic approach of something is better than nothing :) With the approach i have taken, it should be easy to write other strategies for other technologies without having to change the underlying implementation (should just be config change).

**Comment by mike182uk** [comment]
I've done abit more work on my PoC (https://github.com/mike182uk/Ghost/pull/1) and added a strategy for using [search-index](https://www.npmjs.com/package/search-index) instead of lunr (search-index is the one that uses leveldb). `npm install` installed everything necessary to get this working. If you want to test this out, once you have pulled in my latest changes, in `config.js` change `search.strategy` to `search-index`. The db is stored at `content/data/ghost-search-index`. This verifies that the solution is easy to extend for other search technologies :)

I've changed my implementation slightly to now use promises when returning search results, and the search query parsing is now done as part of the pipeline.

**Comment by ErisDS** [comment]
@mike182uk this all sounds amazing - what I can't quite determine from a very quick grok of the code (I need to sit and look at it properly and will try to do that sooner rather than later) is what this gives us in terms of API. Can you provide a quick usage example for the kind of searches that can be done? Thanks!

**Comment by mike182uk** [comment]
Hey @ErisDS 

At the moment i have only added the ability to be able to search posts. You search via a phrase (or keyword). This is done by appending an extra query string param on to the posts retrieval api call:
```
/posts/?search=<phrase_or_keyword>
```

The fields that are used for the search are configureable via the config. In my PoC i have set `title`, `meta_title` and `description` as searchable fields. So if i had the following posts:

title | meta_title | description |
-----|---------------|----------------|
foo | foo            | lorem ipsum doller |
bar | bar           | doller sit amet |
baz | baz         | consectetur adipiscing elit |

and i did the following search:

```
/posts/?search=doller
```

then document `foo` and document `bar` would be returned as they both contain the keyword `doller` in the description.

If i did the search:

```
/posts/?search=consectetur+adipiscing
```

then document `baz` would be returned as it contains the phrase `consectetur adipiscing` in the description.

To give a bit of background on my implementation when you take a look at the code:

- different search technologies are split into different strategies
- a strategy is called a strategy because the strategy pattern is being used
- each strategy is its own self contained unit / module inside of `core/server`
- each strategy follows the naming convention of `search-<technology>` (i.e `search-lunr` for LunrJS) 
- each strategy will contain a `strategy.js` that implements the required methods
- each strategy will contain an indexer init file for each model. This init file deals with initializing the index for that model using that strategy `search-<technology>/model/<model>.js`. So `search-lunr/model/post.js` will initialize the post index for the lunr strategy
- the strategy that is being used is set in the config
- the config allows searchable fields to be defined per model that will be passed to the strategy (at the moment my implementation only deals with posts model, but this could easily be extended)
- the search strategy is abstracted away by `server/core/search`. Ghost uses the objects / functions exposed by this for dealing with search
- the search module provides a `search model resolver` and a `index model resolver`. `index model resolver` will resolve the correct indexer for a model (based on the passed model and the defined search strategy in the config). `search model resolver` will resolve the correct model for searching  (based on the passed model and the defined search strategy in the config). This is how the search strategy can be dynamically changed.

Hopefully that makes sense, but if not let me know :smile:

**Comment by mike182uk** [comment]
@ErisDS did the above clarify things?

**Comment by fergiemcdowall** [comment]
Disclaimer: Maintainer of [search-index] here :)

Its great that a PoC has been made that includes search-index (and a few other cool search technologies). Search-index is a great choice for Ghost for the following reasons:

**Pros**

 * It has been tested and developed on smallish (0-50000) sets of articles, and has many of the features that make it blog-centric: filtering and faceting on metadata (allows word clouds, categorisation, filter by author, etc); autosuggest, and some other cool stuff
 * It exists firmly in the npm ecosystem
 * The scaling up case is nicely handled by the `level` backend: you can run it on any data store you want: text/javascript/indexDB/redis or any other [leveldown backend](https://github.com/Level/levelup/wiki/Modules#storage-back-ends)

**Cons**

 * The query language is still undergoing development. For example: "OR" queries could be handled better.
 * Synonym functionality is possible, but clumsy
 * The default matcher (autocomplete, autosuggest) does not currently support fuzzy matching

There has been a discussion above about lunr.js vs levi vs search-index. They are all equally as good, depending on your use case. I would very cautiously say that lunr.js is trying to be as lucene-like as possible, and that levi was originally developed for a "firehose" user case (fast, streaming indexing).

**Comment by fergiemcdowall** [comment]
Oh, and search-index [is published under the MIT licence](https://github.com/fergiemcdowall/search-index/blob/master/LICENSE)

**Comment by mike182uk** [comment]
@ErisDS Did you have any thoughts on this? I've just updated my fork to work with Ghost 0.7.6

**Comment by gergelyke** [comment]
Recently we added algolia to Ghost, you can take a look at the working example here: https://blog.risingstack.com

I know it is not a free tool, but if you feel like it can be added to make life easier I am happy to help.

**Comment by disordinary** [comment]
One of the things about levelup is that it can have any datastore behind it. 

So I thought I'd test an implementation, I modified ghost to add search functionality and also hacked a version of search-index so that it was backed by sqlite (using sqldown).

I ran two versions of the search-index module side by side within the same blog - one leveldb and the other sqlite backed. I created a blog with fifty articles of random gibberish at approximately 3000 characters each. I then ran the artillery load tester at each search engine for two minutes and found the following:

The **LevelDB** backed search managed to process **5915** requests in two minutes with no failures.

The SQLite backed search managed to process **2363** requests in two minutes and had an additional **889** timeouts. 

So obviously LevelDB is quite a lot faster than an SQLite backed search engine. However, under limited load (which realistically most blogs are) there is no perceptible difference for the end user.

It seems feasible that the stock option for ghost would be to use search-index but store its index in the blogs own database. This seems better, and at least more consistent, than using the FTS functions of the various SQL data stores, and means that upgrading performance when required is as simple as changing a node_module to leveldown.

**Comment by fergiemcdowall** [comment]
@disordinary Thats an interesting benchmark. If SQLite is already a part of the Ghost install bundle, and is going to be so for the foreseeable future- then `search-index` + `SQLite` would seem to be a pretty watertight solution.

**Comment by disordinary** [comment]
@fergiemcdowall interestingly I hadn't saved the reports last time so I ran the load test again, this time SQLite was pretty much on par (in fact its slightly ahead) with levelDB. 

The main difference is previously the version of search-index for leveldb was from npm and the version for sqlite was from github, whereas both are identical. You will know better than I if that would make a difference, my guess is that in perfect conditions sqlite and leveldb are similar but sqlite is more susceptible to things like diskio.

Anyway the branch is here: https://github.com/disordinary/Ghost/tree/search

The latest run of the SQL report is here: https://rawcdn.githack.com/disordinary/Ghost/search/artillery_report_sql.html

The latest run of the LDB report is here: https://rawcdn.githack.com/disordinary/Ghost/search/artillery_report_ldb.html

The signs are promising anyway.

**Comment by jberkus** [comment]
What about testing them against MySQL and PostgreSQL Full Text Search times?  I'd expect these would be significantly faster, but haven't tested.

Adding LevelDB as a dependency to Ghost seems non-trivial.  Also, if you're saying that LevelDB search-index is not dependant on disk, does that mean it's not flushing to disk? What happens on unexpected restart of the VM?  If LevelDB is in-memory only, we'd have to have plans to rebuild the index.

**Comment by disordinary** [comment]
LevelDB takes its name from its architecture, it keeps data in levels and the older the data the more optimised it is for seeks. In hindsight the reason for speed parity is probably that both databases are keeping the entire dataset in memory, further testing would need to be done in reduced in a purpose built vm or on the free tier of amazon and with a larger dataset.

@jberkus what I wast testing above is a search-index based search system that **does not have any dependencies on leveldb,** in the test it has a dependency on SQLite as it's bundled with ghost, however it can also store the indexes in MySQL or Postgre although if those servers are on different systems I'd imagine latency would kill performance but it needs to be tested. 

The main goal was to see if search would be of an acceptable speed without leveldb or FTS and from these limited and unscientific tests it would appear so.

The benefits that search-index has over FTS are:
* It provides consistent search capability regardless of the datastore provided, I know that Ghost is in fact reducing official support to just MySQL and SQLite, however I also know that people are working on removing knex as a dependency of bookshelf and in the future there could be official or unofficial support for nosql databases which could flow on to ghost. 
* If stored in a separate sqlite db (although as I said it can be stored in the production db as well) then we reduce the chance of search having an impact on retrieving blog data.
* Bookshelf and Knex actually have no support for Full Text Search and so there will be a need to work around the abstraction layer.

Of course we need to run benchmarks against FTS as well as storing the search-index index in mysql, and do so with a much larger dataset and more restrictions.

**Comment by jberkus** [comment]
Ah, ok, I misunderstood the relationship of components.  Carry on!

**Comment by disordinary** [comment]
Okay, for the fun of it I also built an FTS index in SQLite using FTS4, it's important to note that FTS isn't compiled into SQLite by default, although the version that is distributed via NPM has it - at least on my platform.

So anyway I ran new tests with an LDB backed index-search, an SQLite backed index-search, and SQLite FTS.

Frustratingly they all performed about the same at just under 6,000 requests served over the two minute test. I ran a few tests and results were consistent. index-search on sqlite just had the edge over FTS on sqlite which was just ahead of index-search on leveldb. But we're talking a difference of half a percent.

As I said more tests and at a larger scale are needed to make a decision, but from these small samples it appears as though it will just come down to features between FTS and search-index. It also shows what a good job @fergiemcdowall  and his team have done.

**Comment by jberkus** [comment]
Sounds good, then.  As previously noted, most users don't have enough posts to really care about FTS features ...

**Comment by fergiemcdowall** [comment]
@disordinary just to confirm: do your benchmarks show that `search-index` with sqlite is slightly faster than `search-index` with leveldb? And was this for indexing or searching?

**Comment by disordinary** [comment]
@fergiemcdowall,

Yes, although there are severe caveats with those figures. I didn't benchmark inserts as in this situation reading is always going to be more important than writing. 

In hindsight each db was so close that the bottleneck is most likely be Ghost, Express and Nginx and not whatever system is hosting the inverted index. 

The dataset and use case is so small though (something like 22,000 records all up from memory) that it could be that SQLite is keeping the entire lot in memory and that as it grows LevelDB will outperform it, this is the likely scenario in my mind.

This experiment did peak my interest though and I've made a start at a more comprehensive test, however life and work got in the way and I haven't yet had a chance to finish it. I'll try and get something done today. Although to really test it we have to push it far beyond the realms of an average blog and into those top 1% of sites that have multiple writers and tens of thousands of readers. Those sites wouldn't be using search-index but rather an elastic search cluster.

**Comment by mlbrgl** [comment]
Adding https://github.com/mlbrgl/ghost-algolia to the discussion. This project aims at adding search to the front-end using a third-party service ([Algolia](https://www.algolia.com)).

_At the time of writing, this project is mainly geared towards developers as it still relies on a core hack to allow indexing. This should be resolved once Ghost apps reach a more mature state._

**Comment by pascalandy** [comment]
I guess the Ghost team is looking for an open-source solution. Algolia looks great, but you hands over your data to them.

**Comment by mlbrgl** [comment]
@pascalandy that's my understanding too :) So you are quite right, this is not necessarily **the** solution for frontend and admin search in Ghost, but still could help some folks getting the ball rolling in my opinion.

**Comment by fergiemcdowall** [comment]
It should be possible to throw together a pretty decent solution with `search-index`- it runs on embedded databases such as `leveldb` (default) and `sqlite3` which I believe is what Ghost uses.

**Comment by jberkus** [comment]
There are any number of solutions that *could* work, the main problem is that nobody is writing the code.

**Comment by zicklag** [comment]
Hey everybody, I haven't had time to read the whole thread yet, but I wanted to mention [Taiga Tribe](https://tribe.taiga.io/) as a possible way to find a developer to work on this. You could make it a $0 gig just to get extra exposure to somebody who might want to do it. The Taiga Tribe community is still pretty limited, but you never know.

**Comment by olivernn** [comment]
Just an FYI, since I was originally involved in this discussion there have been some fairly major [updates](https://lunrjs.com/guides/upgrading.html) to Lunr, probably the most pertinent ones are that indexes are immutable and the serialised index size should be smaller and searches quicker.

As before, I'm happy to help with any Lunr related questions or implementations.

**Comment by paambaati** [comment]
Just in case someone landed here looking for a quick-and-dirty solution, I've written about it here — https://httgp.com/adding-search-to-ghost/

**tl;dr**
Fetch all post content on page load via ServiceWorker and build search index via [Fuse.js](http://fusejs.io/).

**Comment by pascalandy** [comment]
@paambaati Do you have a PR somewhere with the solution's implementation you are describing in your post?

**Comment by paambaati** [comment]
@pascalandy Not really, because the solution will work for only small blogs, and wouldn't scale for larger blogs. This discussion brings up some other interesting and yet valid use cases, and hence the lack of a PR.

**Comment by kevinansfield** [comment]
@pascalandy, @paambaati this is essentially what [GhostHunter](https://github.com/jamalneufeld/ghostHunter) (most common Ghost search plugin) does except that uses lunr.js instead of fuse.js. It does scale to large blogs (100s of posts) reasonably well but very large blogs (1000s of posts) can struggle, it requires a fair bit of server-side processing to generate the API response and then you have the download time as well although that generally won't me much bigger than a large image.

There is also an [algolia integration](https://github.com/mlbrgl/ghost-algolia) but bear in mind it currently requires core to be patched every time you upgrade.

We have an [Ideas topic](https://forum.ghost.org/c/Ideas/l/votes) on our forum now, it's worth adding your vote to the [Search](https://forum.ghost.org/t/search-for-ghost/832) idea and joining the discussion there.

**Comment by jozic** [comment]
I use twitter's typeahed, it's far from ideal, but works
see it live here https://onlydarkchocolate.com/
and here's PR to my Casper fork https://github.com/jozic/Casper/commit/bde7c790cae9a69a4c5f72099f8d9749a9597644

**Comment by pascalandy** [comment]
It works very well @jozic. What makes it far from ideal?
Thanks for the PR !!

**Comment by jozic** [comment]
@pascalandy 
things i don't like
- when you select an item from the list, before opening the post it puts selected item as json in search field (minor)
- sometimes it will not open an item, not very often, but it's not great experience for sure

also it loads all posts and does search on client side, which won't scale (but I'm not at that scale yet)

otherwise, it works fine

**Comment by andy1247008998** [comment]
3 years passed

**Comment by vikaspotluri123** [comment]
@andy1247008998 The same sentiment in https://github.com/TryGhost/Ghost/issues/4453#issuecomment-387252241 applies here.

**Comment by juan-g** [comment]
This feature can be voted at [Search for Ghost - Ideas - Ghost Forum](https://forum.ghost.org/t/search-for-ghost/832), as Kevin suggested.

**Comment by lengerrong** [comment]
just an idea.

use sqlite3 as database for your ghost.
then there will be a database file.
open the database file via sqlite3.

$sqlite3  content/data/ghost-dev.db      
SQLite version 3.11.0 2016-02-15 17:29:24
Enter ".help" for usage hints.
sqlite> .tables
accesstokens            migrations              refreshtokens         
api_keys                migrations_lock         roles                 
app_fields              mobiledoc_revisions     roles_users           
app_settings            permissions             sessions              
apps                    permissions_apps        settings              
brute                   permissions_roles       subscribers           
client_trusted_domains  permissions_users       tags                  
clients                 posts                   users                 
integrations            posts_authors           webhooks              
invites                 posts_tags            

As you can see, all posts definitely stored in the table named 'posts'.

sqlite> select * from posts;
id|uuid|title|slug|mobiledoc|html|comment_id|plaintext|feature_image|featured|page|status|locale|visibility|meta_title|meta_description|author_id|created_at|created_by|updated_at|updated_by|published_at|published_by|custom_excerpt|codeinjection_head|codeinjection_foot|og_image|og_title|og_description|twitter_image|twitter_title|twitter_description|custom_template

Then we got the structure of the 'posts' table.
As I tested, 'slug' is the route path of a recorded post.

let's search posts contains 'welcome' as an example.
sqlite> select slug from posts where html like '%welcome%';
slug
apps-integrations
welcome

then we can check the real post in browser.
http://localhost:2368/welcome
http://localhost:2368/apps-integrations

see, search things goes well in a shell bash.
all we need to do write js codes to do above operations and display the result correctly in ghost core server module.

anyone who familiar with ghost developing work can have a try for that.
any suggestion please guide.
I would like to implement it at part time.
Hope I can finish it soon.

**Comment by zicklag** [comment]
[GhostSearch](https://github.com/HauntedThemes/ghost-search) by HauntedThemes looks like it works for my small blog.

**Comment by Windyo** [comment]
Quick ping here : does it make sense to continue maintaining GhostHunter ? With the V2 having been just released, I'm wondering if you guys are closer to releasing the Apps API, or maybe your own search, or if I should keep maintaining the package.

**Comment by ErisDS** [comment]
@Windyo we are not really closer to implementing search or any sort of apps into Ghost itself, instead we are focusing on providing extensibility through the API & Webhooks. 

The new v2 API will offer two relevant things: 

1) an easier integration point with something like Algolia through webhooks, although it will still require a separate service to be running to, for example process the webhooks & send data to algolia in the right format
2) the v2 API is designed to be cached, which means that it is much better at handling the kind of heavy usage that busy sites might run through GhostHunter.

GhostHunter is currently one of the most commonly used extensions for Ghost and I think it will stay that way for a little while longer yet 😊

**Comment by Windyo** [comment]
Alright @ErisDS I'll keep GH up to date at least then - May have ideas for minor improvements yet.
Let me know if the latest changes make it better for your server load.

**Comment by r3wt** [comment]
I am building a route in our node.js proxy today which will query the ghost api at startup, build an index, and then periodically query and update the index. this will allow me to add search capabilities for our ghost blog, which is headless (we have a react webapp, and only use ghost api). when i finish, i'll share the code for you here, and someone can port it into ghost if the team wants.

**Comment by r3wt** [comment]
as promised, i've written a basic implementation of a search engine for ghost. now, this utilizes the `tryghost/content-api` library for fetching the posts, and it runs as a route in express app. i was hoping to throw up a gist for this, but working in a startup and the need to move fast, i really don't have time. i'm just gonna drop this mvp right here for those who might wonder how to do this easily. 

**GhostSearch.js**
```js
// begin config
const ghost_host = '<your host here>';//your blogs domain -- see documentation of tryghost/content-api
const ghost_key = '<your key here>';//key to your integration -- see documentation of tryghost/content-api
const updateInterval = 60000;//how often the index should update given in milliseconds
const requestTimeout = 5000;// maximum time for a search request before it times out given in milliseconds (note this isn't exact, it depends on the load of the event loop)
const resultsPerPage = [8];// allowed values for limit. in my case we only allow 8, but made this configurable so users of ghost can benefit

// end config

const elasticlunr = require('elasticlunr');

var index = elasticlunr(function () {
    this.addField('title');
    this.addField('meta_title');
    this.addField('plaintext');
    this.addField('meta_description');
    this.addField('excerpt');
    this.addField('custom_excerpt');
    this.setRef('id');
});

const GhostContentAPI = require('@tryghost/content-api');

const blog = new GhostContentAPI({
    host: ghost_host,
    key: ghost_key,
    version: 'v2'
});

var inprogress = 0,
    indexWillUpdate = false,
    indexReady = false

function updateIndex() {
    return new Promise(updateIndex_impl);//kicks of the update by creating a promise and calling the internal updateIndex_impl
}

// this function accepts the resolve/reject from the promise and implements the logic for waiting until it can update the index, then doing so 
// don't have to worry about uncaught promise errors, we just keep recursing
function updateIndex_impl(resolve,reject) {
    if(inprogress > 0) {
        setTimeout(()=>updateIndex(resolve,reject),22);//just keep passing the original arguments
    }else{
        indexWillUpdate = true;
        console.log('updating index')
        blog.posts.browse({limit:'all',formats:'html,plaintext'}).then(posts=>{
            console.log('got posts from blog');
            posts.forEach(post=>index.addDoc(post));//If docRef already exist, then update doc. <--taken straight from elasticlunr docs
            indexWillUpdate=false;
            indexReady = true;
            console.log('index updated');
            resolve();
        })
        .catch(err=>{
            console.log('failed to update index');
            console.log(err);
            reject(err);
        })
    }
}

function waitForIndex(timeout) {
    return new Promise((resolve,reject)=>{
        if(indexWillUpdate || !indexReady){
            var elapsed = 0;
            var _t = setInterval(()=>{
                elapsed += 66;
                if(!indexWillUpdate && indexReady){
                    clearInterval(_t);
                    resolve();
                }else{
                    if(elapsed > timeout){
                        clearInterval(_t);
                        reject(new Error('timeout exceeded while waiting for index'));
                    }  
                }
                
            },66);
        }else{
            resolve();
        }
    })
}

function search(req,res) {

    const term = req.query.term || false;
    const limit = req.query.limit ? ~~(req.query.limit) : resultsPerPage[0];
    const page = (req.query.page ? ~~(req.query.page) : 1) - 1;

    if(!term){
        return res.status(400).json({posts:null,message:'missing required parameter `term`'});
    }

    if(resultsPerPage.indexOf(limit) === -1){
        return res.status(400).json({posts:null,message:'supplied value for parameter `limit` is invalid'});
    }

    if(page < 0){
        return res.status(400).json({posts:null,message:'supplied vlaue for parameter `page` is invalid'});
    }

    waitForIndex(requestTimeout).then(()=>{
        inprogress +=1;
        
        var startIndex = page * limit;
        var endIndex = (page + 1) * limit;

        var posts = index.search(term,{
            fields: {
                title: { boost: 2, bool: 'AND' },
                plaintext: { boost: 1, bool: 'AND' }
            },
            bool:'OR',
            expand: true
        }).map(item=>index.documentStore.getDoc(item.ref)).slice(startIndex,endIndex);
        res.status(200).json({ posts });
        inprogress -= 1;
    })
    .catch(err=>{
        console.log(err);
        res.status(500).json({posts:null,message:'500 internal server error'});
    });
}

//kick it off
function routine(){
    updateIndex().then(()=>setTimeout(routine,updateInterval)).catch(err=>{
        console.log(err);
    })
}

// in my version we just kick it off here from the module, but to make it more startup friendly this module exports a function to manually start the engine
// routine();//starts the index updating routine

const run = ()=>routine();

module.exports = {search,run};
```

**suggested usage in an express app**
```js
const app = require('express')();
const GhostSearch = require('/path/to/GhostSearch.js');
GhostSearch.run();
app.get('/search',GhostSearch.search);
```

I'll try and update this later as i have time. hope you all can find some use from this.

**Comment by vikaspotluri123** [comment]
@r3wt the implementation seems similar to [jamalneufeld/ghostHunter](https://github.com/jamalneufeld/ghostHunter) and the [serverside implementation](https://github.com/hexrweb/ghosthunter-server) (I've done quite a bit of work on the serverside version). There's also [HauntedThemes/ghost-search](https://github.com/HauntedThemes/ghost-search) which is a frontend-only implementation

**Comment by r3wt** [comment]
@vikaspotluri123 not sure why you would say that, but ok. my implementation uses `elasticlunr` rather than `lunr`, and also works with the v2 ghost api, leveraging the `content-api` package from GhostSDK. The only real similarities i see are that they provide a simple search endpoint for use with a ghost blog. I think the advantage of mine is that you make no changes to ghost. you just go into the admin interface and add an integration and grab the key. pretty simple. the code above is unlicensed, anyone can have it and freely modify it. 

best of luck.

r3

**Comment by vikaspotluri123** [comment]
@r3wt great points! GH-Server hasn't been updated for the V2 API, but basically does the same thing, just using the v0.1 public API 😉 GH-Server is also designed to run in parallel to Ghost, and also has frontend integration because GH does 😂 I don't want to pollute this thread, so if you want to talk specifics or collaborate, feel free to drop a line 😊

**Comment by r3wt** [comment]
@vikaspotluri123 Sounds cool, if i take a larger interest in ghost i will definitely take you up on on that. Thanks

**Comment by andy1247008998** [comment]
This issue should not be closed until solved.

**Comment by danieldanielecki** [comment]
+1
